File: packages/go/pkg/bot/ephemeral.go
package bot

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"

	"github.com/Nam088/telebot/packages/go/pkg/types"
)

// EditEphemeralMessageText edits text or rich content of an ephemeral message
// (Bot API 10.2+).
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - opts: Options identifying the ephemeral message plus the new text or rich content.
//
// Returns:
//   - *types.Message: The edited Message, non-nil when Telegram echoes one.
//   - bool: True when Telegram returned the bare boolean result instead of a Message.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	msg, ok, err := b.EditEphemeralMessageText(ctx, &types.EditEphemeralMessageTextOptions{
//		ChatID:             int64(-1001234567890),
//		ReceiverUserID:     123456,
//		EphemeralMessageID: 7,
//		Text:               "Corrected text",
//	})
func (b *Bot) EditEphemeralMessageText(ctx context.Context, opts *types.EditEphemeralMessageTextOptions) (*types.Message, bool, error) {
	return b.requestMessageOrTrue(ctx, "editEphemeralMessageText", opts)
}

// EditEphemeralMessageMedia edits the media of an ephemeral message (Bot API 10.3+).
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - opts: Options identifying the ephemeral message plus the new media object.
//
// Returns:
//   - *types.Message: The edited Message, non-nil when Telegram echoes one.
//   - bool: True when Telegram returned the bare boolean result instead of a Message.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	msg, ok, err := b.EditEphemeralMessageMedia(ctx, &types.EditEphemeralMessageMediaOptions{
//		ChatID:             int64(-1001234567890),
//		ReceiverUserID:     123456,
//		EphemeralMessageID: 7,
//		Media:              &types.InputMediaPhoto{Type: "photo", Media: "AGACAD..."},
//	})
func (b *Bot) EditEphemeralMessageMedia(ctx context.Context, opts *types.EditEphemeralMessageMediaOptions) (*types.Message, bool, error) {
	return b.requestMessageOrTrue(ctx, "editEphemeralMessageMedia", opts)
}

// EditEphemeralMessageCaption edits the caption of an ephemeral message
// (Bot API 10.3+).
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - opts: Options identifying the ephemeral message plus the new caption.
//
// Returns:
//   - *types.Message: The edited Message, non-nil when Telegram echoes one.
//   - bool: True when Telegram returned the bare boolean result instead of a Message.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	msg, ok, err := b.EditEphemeralMessageCaption(ctx, &types.EditEphemeralMessageCaptionOptions{
//		ChatID:             int64(-1001234567890),
//		ReceiverUserID:     123456,
//		EphemeralMessageID: 7,
//		Caption:            "New caption",
//	})
func (b *Bot) EditEphemeralMessageCaption(ctx context.Context, opts *types.EditEphemeralMessageCaptionOptions) (*types.Message, bool, error) {
	return b.requestMessageOrTrue(ctx, "editEphemeralMessageCaption", opts)
}

// EditEphemeralMessageReplyMarkup edits the inline keyboard attached to an
// ephemeral message (Bot API 10.3+).
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - opts: Options identifying the ephemeral message plus the new reply markup.
//
// Returns:
//   - *types.Message: The edited Message, non-nil when Telegram echoes one.
//   - bool: True when Telegram returned the bare boolean result instead of a Message.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	msg, ok, err := b.EditEphemeralMessageReplyMarkup(ctx, &types.EditEphemeralMessageReplyMarkupOptions{
//		ChatID:             int64(-1001234567890),
//		ReceiverUserID:     123456,
//		EphemeralMessageID: 7,
//	})
func (b *Bot) EditEphemeralMessageReplyMarkup(ctx context.Context, opts *types.EditEphemeralMessageReplyMarkupOptions) (*types.Message, bool, error) {
	return b.requestMessageOrTrue(ctx, "editEphemeralMessageReplyMarkup", opts)
}

// DeleteEphemeralMessage deletes an ephemeral message (Bot API 10.2+).
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - opts: Options carrying chat_id, receiver_user_id and ephemeral_message_id.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.DeleteEphemeralMessage(ctx, &types.DeleteEphemeralMessageOptions{
//		ChatID:             int64(-1001234567890),
//		ReceiverUserID:     123456,
//		EphemeralMessageID: 7,
//	})
func (b *Bot) DeleteEphemeralMessage(ctx context.Context, opts *types.DeleteEphemeralMessageOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "deleteEphemeralMessage", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// requestMessageOrTrue sends a Bot API method whose result is the union
// "Message | True" (node types it as `Message | boolean`) and splits the two
// possible wire shapes into separate Go values.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - method: Bot API method name.
//   - payload: Request payload; may be nil.
//
// Returns:
//   - *types.Message: The edited Message, non-nil only when Telegram echoes one.
//   - bool: True when Telegram returned the bare boolean result.
//   - error: TelegramError or a decoding error.
func (b *Bot) requestMessageOrTrue(ctx context.Context, method string, payload any) (*types.Message, bool, error) {
	var raw json.RawMessage
	if err := b.Request(ctx, method, payload, &raw); err != nil {
		return nil, false, err
	}
	trimmed := bytes.TrimLeft(raw, " \t\r\n")
	if len(trimmed) > 0 && (trimmed[0] == 't' || trimmed[0] == 'f') {
		var ok bool
		if err := json.Unmarshal(raw, &ok); err != nil {
			return nil, false, fmt.Errorf("failed to unmarshal %s boolean result: %w", method, err)
		}
		return nil, ok, nil
	}
	var msg types.Message
	if err := json.Unmarshal(raw, &msg); err != nil {
		return nil, false, fmt.Errorf("failed to unmarshal %s message result: %w", method, err)
	}
	return &msg, false, nil
}

File: packages/node/src/client/methods/business/ephemeral.ts
/**
 * Rich message, ephemeral messages, and request methods for Bot API.
 *
 * @packageDocumentation
 */

import { BusinessGiftsMethods } from "./gifts.js";
import type {
  Message,
  SendRichMessageOptions,
  SendRichMessageDraftOptions,
  EditEphemeralMessageTextOptions,
  EditEphemeralMessageMediaOptions,
  EditEphemeralMessageCaptionOptions,
  EditEphemeralMessageReplyMarkupOptions,
  DeleteEphemeralMessageOptions,
  AnswerChatJoinRequestQueryOptions,
  SendChatJoinRequestWebAppOptions,
} from "../../types/index.js";

/**
 * Mixin providing rich messages, ephemeral messages, and low-level API operations.
 */
export abstract class BusinessEphemeralMethods extends BusinessGiftsMethods {
  /**
   * Sends a structured rich formatted message (Bot API 10.1+).
   *
   * @param options - Rich message options including `chat_id` and `rich_message`.
   * @returns The sent {@link Message}.
   *
   * @see {@link https://core.telegram.org/bots/api#sendrichmessage Telegram Bot API: sendRichMessage}
   */
  public async sendRichMessage(options: SendRichMessageOptions): Promise<Message> {
    return this.request<Message>("sendRichMessage", options as unknown as Record<string, unknown>);
  }

  /**
   * Streams a draft of a rich formatted message (Bot API 10.1+).
   *
   * @param options - Rich message draft options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#sendrichmessagedraft Telegram Bot API: sendRichMessageDraft}
   */
  public async sendRichMessageDraft(options: SendRichMessageDraftOptions): Promise<boolean> {
    return this.request<boolean>(
      "sendRichMessageDraft",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Edits an ephemeral message text (Bot API 10.2+).
   *
   * @param options - Ephemeral message edit options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#editephemeralmessagetext Telegram Bot API: editEphemeralMessageText}
   */
  public async editEphemeralMessageText(
    options: EditEphemeralMessageTextOptions,
  ): Promise<boolean> {
    return this.request<boolean>(
      "editEphemeralMessageText",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Edits an ephemeral message media (Bot API 10.3+).
   *
   * @param options - Ephemeral message media edit options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#editephemeralmessagemedia Telegram Bot API: editEphemeralMessageMedia}
   */
  public async editEphemeralMessageMedia(
    options: EditEphemeralMessageMediaOptions,
  ): Promise<boolean> {
    return this.request<boolean>(
      "editEphemeralMessageMedia",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Edits an ephemeral message caption (Bot API 10.3+).
   *
   * @param options - Ephemeral message caption edit options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#editephemeralmessagecaption Telegram Bot API: editEphemeralMessageCaption}
   */
  public async editEphemeralMessageCaption(
    options: EditEphemeralMessageCaptionOptions,
  ): Promise<boolean> {
    return this.request<boolean>(
      "editEphemeralMessageCaption",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Edits an ephemeral message reply markup (Bot API 10.3+).
   *
   * @param options - Ephemeral message reply markup edit options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#editephemeralmessagereplymarkup Telegram Bot API: editEphemeralMessageReplyMarkup}
   */
  public async editEphemeralMessageReplyMarkup(
    options: EditEphemeralMessageReplyMarkupOptions,
  ): Promise<boolean> {
    return this.request<boolean>(
      "editEphemeralMessageReplyMarkup",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Deletes an ephemeral message (Bot API 10.2+).
   *
   * @param optionsOrChatId - Options object or chat identifier.
   * @param receiverUserId - Receiver user identifier when passing positional arguments.
   * @param ephemeralMessageId - Ephemeral message identifier when passing positional arguments.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#deleteephemeralmessage Telegram Bot API: deleteEphemeralMessage}
   */
  public async deleteEphemeralMessage(
    optionsOrChatId: DeleteEphemeralMessageOptions | number | string,
    receiverUserId?: number,
    ephemeralMessageId?: number,
  ): Promise<boolean> {
    if (typeof optionsOrChatId === "object") {
      return this.request<boolean>(
        "deleteEphemeralMessage",
        optionsOrChatId as unknown as Record<string, unknown>,
      );
    }
    return this.request<boolean>("deleteEphemeralMessage", {
      chat_id: optionsOrChatId,
      receiver_user_id: receiverUserId,
      ephemeral_message_id: ephemeralMessageId,
    });
  }

  /**
   * Answers a chat join request query from a user (Bot API 10.1+).
   *
   * @param options - Query response options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#answerchatjoinrequestquery Telegram Bot API: answerChatJoinRequestQuery}
   */
  public async answerChatJoinRequestQuery(
    options: AnswerChatJoinRequestQueryOptions,
  ): Promise<boolean> {
    return this.request<boolean>(
      "answerChatJoinRequestQuery",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Sends a Web App for a chat join request (Bot API 10.1+).
   *
   * @param options - Join request Web App options.
   * @returns `true` on success.
   *
   * @see {@link https://core.telegram.org/bots/api#sendchatjoinrequestwebapp Telegram Bot API: sendChatJoinRequestWebApp}
   */
  public async sendChatJoinRequestWebApp(
    options: SendChatJoinRequestWebAppOptions,
  ): Promise<boolean> {
    return this.request<boolean>(
      "sendChatJoinRequestWebApp",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * General-purpose raw API method executor.
   *
   * @typeParam T - Expected result payload type.
   * @param method - API method name.
   * @param payload - Request parameters.
   * @returns Unwrapped API result.
   */
  public async doApiRequest<T>(method: string, payload: Record<string, unknown> = {}): Promise<T> {
    return this.request<T>(method, payload);
  }
}
